### ISSUE TITLE
mAP calculation inflated (and also affected by the conf parameter)

### ISSUE THREAD

**Issue by vittorio-prodomo** [ultralytics/ultralytics#20970]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

I want to clarify that I only tested segmentation yolo model v11, and only for validation, but what I discovered sounds like something that would happen in any mode (train, val) and with any model (seg or detection).

When running 'yolo segment val', we can explicitly set the conf threshold value. In truth, it sounds like this parameter would only be useful for a predict and not for the validation script, since most of the results provided by Ultralytics are results based on **doing a sweep of different threshold values**. Anyway, even assuming that the presence of this parameter in the model.val() signature is intended behavior, and it only affects the scalar numbers returned for Precision and Recall (the only place of the results where I reckon it would make sense), that's still not what happens. Basically, when I explicitly set the confidence threshold to something higher than the default (0.001), then the detections are **seemingly first and foremost filtered according to that threshold**, before doing anything else. Then the usual workflow happens: the remaining predicted labels and the ground truth ones are compared based on a IoU threshold to obtain our values of TP, FP and FN, and the Precision (P) and Recall (R) curves are computed from those by letting the conf threshold vary between 0 and 1 with micro steps.

From the P and R curves, the mAP curve is obtained. So what I noticed is that if we try to increase this confidence threshold and we re-launch the validation, we observe in the P and R curves an increasing portion of the first part of the curve that becomes constant. This makes sense if the intended behavior is this of a weird "pre-filtering" of detections before the confidence sweep, because it means (if you think about it in reverse, going from right to left) that once you go lower than a certain conf, the number of detections doesn't change (since those detections have been dropped) so your P and R values stay the same.

Now, another problem arises in the way the mAP curve is actually computed and drawn. One would expect (correct me if I am wrong) that the true, theoretical mAP curve simply "ends" at the last point (the point that represents the P and R values of the model when the lowest confidence threshold possible is tested). This means that a vertical line, projected straight from that point to the x-axis of the mAP plot should represent the true end of the curve, visually speaking. Instead, Ultralytics artificially draws the rest of the curve going from that last point to the fixed, bottom-right corner of the plot. Since the plot is fixed to be limited between 0 and 1 on both axes, it means you will notice this straight line (but not vertical) that goes from the last point to the (0,1) point. This means there is a whole new portion "under the curve" that will affect the mAP value.

So, going back to our threshold value, assuming this is intended behavior (which I am not sure about) and we are supposed to be able to use it to pre-filter the detections before the whole sweep on confidence thresholds happens, then: if we try to increase it, we should see the mAP curve more and more truncated on the right (because after a certain point, all the points plotted on the mAP plot coincide, since neither P nor R are changing), which should translate into a lower mAP, **always**. Instead, because of the phenomenon I just described, we can actually see the mAP **increasing** (depending on the shape of our curve). Of course this is more and more noticeable depending on how far from the bottom-right corner the last point of the curve is. If, for some reason, at super low confidence your model still doesn't reach values of Precision and Recall that are close to, respectively, 0 and 1, then this phenomenon is visible even with default settings (i.e., conf threshold equal to 0.001).

Let me show you the plots: running two validations, where one uses a default conf of 0.001 and the other uses 0.5, I obtain these two plots:

Plot with 0.5 conf:

![Image](https://github.com/user-attachments/assets/3473689a-c649-4be0-bc7b-6e9ff9557c3f)

Plot with 0.001 conf (where I added two black lines to point where the area should end and also where the orange line goes in the previous plot):

![Image](https://github.com/user-attachments/assets/7c1c0d9e-ccba-428f-b287-7f5c05f2b00d)

As you can see, in my specific case it was even easier to spot this problem because I noticed my mAP numbers were increasing after increasing the conf threshold (which, again, doesn't make sense mathematically speaking). This happens because my orange class has very poor performance so the mAP curve is more "L-shaped" instead of having the typical “hump” (with the center of the curve much closer to the top-right corner) of models that perform well. This translates into what you saw in my plots, where a sudden interruption in my orange curve and a straight line connecting it to the bottom-right corner actually equals to more area under the curve. But if we had a "hump", it would be the opposite because we would be "truncating" part of that hump.

Sorry for the long explanation, I hope I was clear enough.

### Environment

```
Ultralytics 8.3.122 🚀 Python-3.10.12 torch-2.6.0+cu124 CUDA:0 (NVIDIA RTX 6000 Ada Generation, 48520MiB)
Setup complete ✅ (128 CPUs, 251.2 GB RAM, 190.8/1832.2 GB disk)

OS                  Linux-6.8.0-60-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 251.17 GB
Disk                190.8/1832.2 GB
CPU                 AMD Ryzen Threadripper PRO 7985WX 64-Cores
CPU count           128
GPU                 NVIDIA RTX 6000 Ada Generation, 48520MiB
GPU count           2
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

`yolo segment val data=your-data.yaml model=yolo11x-seg.pt conf=0.5`

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @vittorio-prodomo, thank you for your detailed report and for your interest in Ultralytics 🚀! This is an automated response to help you get started—an Ultralytics engineer will review your issue and assist you further soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for additional context. You’ll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, plus answers to common questions.

Since this is a 🐛 Bug Report, please ensure your [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) is complete and accurate to assist our debugging process.

If you’re working on custom training or have additional questions, please provide more details such as dataset image samples, training logs, and confirm you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community wherever you prefer:
- For real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussion: [Discourse](https://community.ultralytics.com/)
- For community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Make sure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to see if the issue persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/), preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your report and for supporting Ultralytics! 🏆

**Comment by Y-T-G** [comment]
It's a known bug reported here:
https://github.com/ultralytics/ultralytics/issues/17270

I asked the original author to open a PR, but he didn't respond after that. If you're interested, you can open a PR with the fix proposed by him.

**Comment by vittorio-prodomo** [comment]
Oh I'm sorry, I reported this in another related issue and @glenn-jocher told me to open an issue. So I assumed it was not known.

**Comment by glenn-jocher** [comment]
No worries about the confusion! You're right that I suggested opening an issue when you mentioned this in another thread - I should have checked for existing reports first. Your detailed analysis with the plots is actually very helpful and provides excellent additional context to the original issue #17270. 

Since this is a duplicate, we can close this one, but if you're interested in contributing a fix based on the proposed solution in #17270, that would be greatly appreciated by the community!

**Comment by ksv87** [comment]
> It's a known bug reported here: [#17270](https://github.com/ultralytics/ultralytics/issues/17270)
> 
> I asked the original author to open a PR, but he didn't respond after that. If you're interested, you can open a PR with the fix proposed by him.

I did a PR and responded https://github.com/ultralytics/ultralytics/issues/17270#issuecomment-2728314961

**Comment by Y-T-G** [comment]
@ksv87 Ah, okay. Sorry, forgot about that.
